### PR TITLE
Fix spawning of mixed-parentage tasks.

### LINKED ISSUE #4913: Runahead limit consistency: P1 vs PT1H
For datetime cycling we can use a time interval `runahead limit`, where the number of active cycles is determined by `<base-cycle-point> + <time-interval>`.  So for a `PT1H` cycle, a  `PT0H` limit allows 1 point (just the base one) to run at a time, `PT1H` allows the base point and one additional point (i.e. two points in total), and so on.

For Cylc 8 we merged the old `max active cycle points` (number of active cycles) setting into `runahead limit` specified as an integer interval (for both integer and datetime cycling), but it seems we forgot to change the value accordingly and ended up with an off-by-one error:
- `P0` means one active point (the base one) :heavy_check_mark: 
- `P1` means ONE active point (the base one) :heavy_multiplication_x:  (should be TWO: base point + a 1-cycle interval)
- `P2` means TWO active points (the base, plus the next one) :heavy_multiplication_x:  (should be THREE ...)
- etc.

IMO this is a bug. The integer interval limit is a new thing in Cylc 8, so we can fix it as long as we do it quickly.  And document that a null runahead **interval** limit is required to restrict activity to one point at a time.

Assigning myself as this will be easier to do on top of or after #4906

Agreed? https://github.com/cylc/cylc-flow/labels/question 

(This might possibly be controversial if people think `P1 == one cycle` is fine - but IMO we shouldn't be using the interval-from-base-point notation if it doesn't actually mean what it says).

### PR DESCRIPTION
**Pull request by hjoliver**
(Those that have parents in some cycles but not in others).

Before spawn-on-demand, all tasks depended implicitly on their own previous instance submitting (same task at previous cycle point), which had several unpleasant consequences such as the inability of same-task instances to run out of order. 

This fixes the last vestige of previous-instance dependence in Cylc 8: if a task instance has no parents to spawn it, we spawn it automatically when its previous instance is released from runahead. This (correctly) makes parentless tasks spawn out to the runahead limit immediately, **but** only if they have no parents in all cycle points. If they have parents at some points, those points temporarily block further spawning (because the parented instance does not get released from runahead until after it gets spawned by its parent).

This PR also tidies spawning-related code a lot (it was really hard to follow after my incremental shoe-horning of SOD into the old codebase ~2 yrs ago :facepalm: :), and computes the runahead limit much less often.

```ini
[scheduling]
    cycling mode = integer
    initial cycle point = 1
    final cycle point = 10
    [[xtriggers]]
        never = xrandom("0")  # never satisfied
    [[graph]]
        # foo has no parents, should spawn to runahead limit
        P1 = "foo => bar"
        # except at point 3, where it should be spawned by baz
        R1/3 = "@never => baz => foo"
[runtime]
    [[foo, bar, baz]]
        script = sleep 10
```
On master, `foo` should auto-spawn to the runahead limit except for point `3` where it will be spawned by `3/baz`, but spawning is blocked at point `3`:

Screenshot from 2022-06-13 14-39-59

On this branch:
Screenshot from 2022-06-13 14-40-54

Also: close #4913 

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` and `conda-environment.yml`.
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] (master branch) I have opened a documentation PR at https://github.com/cylc/cylc-doc/pull/479 (for the runahead fix)

### CODE REVIEW

File: cylc/flow/id_match.py [hunk 1]
```diff
@@ -145,7 +145,7 @@ def filter_ids(
             if tokens.get(lowest_token.value):
                 break
 
-        cycles = []
+        cycles = set()
```
Comment (hjoliver) [inline cylc/flow/id_match.py:148]: (This branch incorporates the bug fix of #4909).

File: cylc/flow/rundb.py [hunk 2]
```diff
@@ -820,7 +820,8 @@ def select_task_pool_for_restart(self, callback):
             LEFT OUTER JOIN
                 %(task_outputs)s
             ON  %(task_pool)s.cycle == %(task_outputs)s.cycle AND
-                %(task_pool)s.name == %(task_outputs)s.name
+                %(task_pool)s.name == %(task_outputs)s.name AND
+                %(task_pool)s.flow_nums == %(task_outputs)s.flow_nums
```
Comment (hjoliver) [inline cylc/flow/rundb.py:824]: Bug fix!  Loading the task pool from the DB at restart, on master I get multiple rows per task proxy if a waiting task had also run in previous flows (and hence attempt to spawn multiple copies of the same task).

**Comment by hjoliver** [comment]
One issue to deal with:
> tests/f/runahead/05-check-default-future-2.t (Wstat: 0 Tests: 5 Failed: 1)

**Comment by hjoliver** [comment]
(NB: I've just pushed a commit to fix the runahead off-by-one error https://github.com/cylc/cylc-flow/issues/4913; this will require a minor adjustment to a whole bunch of tests... which were all passing)

**Review by dwsutherland — APPROVED** [review]
Reviewed, and found working  (could reproduce the issue and fix) 👍 
I reviewed the code, followed most of it...

**Comment by hjoliver** [comment]
NB: I have one thing to investigate on this branch: it might stop the inactivity timer working properly (if so, should be easy to fix).

**Comment by oliver-sanders** [comment]
@hjoliver What font are you using in those terminal screenshots?

**Comment by hjoliver** [comment]
@MetRonnie  - this is good to go now.

**Comment by MetRonnie** [comment]
I'm happy to review but will be fairly busy this upcoming week with workshops, so someone might like to beat me to it

**Comment by hjoliver** [comment]
That's fine (I just pinged you as you'd self-assigned as a reviewer above).

File: cylc/flow/scheduler.py [hunk 3]
```diff
@@ -1846,19 +1815,19 @@ def check_auto_shutdown(self):
...
             # runahead, then held, queued, or xtriggered).
             return False
 
         # Can shut down.
         if self.pool.stop_point:
-            self.options.stopcp = None
-            self.pool.stop_point = None
+            # Forget early stop point in case of a restart.
             self.workflow_db_mgr.delete_workflow_stop_cycle_point()
-            self.update_data_store()
```
[the remark this answers is not in the document]
Comment (MetRonnie) [reply cylc/flow/scheduler.py:1904]: Looking at blame: https://github.com/cylc/cylc-flow/commit/0dbf8b3c91c2962fa1e204c2d247175e02c21f2b

> Scheduler attributes e.g. `is_paused` and `stop_mode` are reflected in the data store.

**Review by MetRonnie — COMMENTED** [review]
Partial review

File: cylc/flow/task_pool.py [hunk 4]
```diff
@@ -168,12 +163,27 @@ def _swap_out(self, itask):
...
+        for name in self.config.get_task_name_list():
+            tdef = self.config.get_taskdef(name)
+            point = tdef.first_point(self.config.start_point)
+            self.spawn_to_rh_limit(tdef, point, {flow_num}, abs_ok=False)
+
+    def add_to_pool(self, itask, is_new: bool = True) -> None:
+        """Add a task to the hidden (if not satisfied) or main task pool.
+
+        If task already exists in the hidden pool and is satisfied, move it to
+        main. TODO: can this be clarified? (Used to update prereqs via
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/task_pool.py:183]: Just a docstring tweak.

File: cylc/flow/task_pool.py [hunk 5]
```diff
@@ -225,16 +241,11 @@ def create_data_store_elements(self, itask):
...
-        point (and <= the stop point, if there is one).
-
-        Incomplete tasks and partially satisfied prerequisites are counted
-        toward the runahead limit, because they represent tasks that will
-        (or may, in the case of prerequisites) yet run at their cycle points.
+        """Release tasks below the runahead limit.
 
         Return True if any tasks released, else False.
+        Call if RH limit changes.
+        TODO: call from that method?
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/task_pool.py:248]: (Reviewed and decided not to "call from that method").

File: cylc/flow/task_pool.py [hunk 6]
```diff
@@ -267,100 +278,123 @@ def release_runahead_tasks(self):
...
+        The limit itself is limited by workflow stop point, if there is one.
+        And adjusted upward on the fly if tasks with future offsets appear.
+        """
+        points: List['PointBase'] = []
+        if not self.main_pool:
+            # Start at first point in each sequence, after the initial point.
+            pts = set()
+            for seq in self.config.sequences:
+                pts.add(seq.get_first_point(self.config.start_point))
+            points = list(pts)
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/task_pool.py:312]: The `set` was intentional, to avoid duplicates (most sequences likely have the same start point). It should not be a large number of points.

File: cylc/flow/task_pool.py [hunk 7]
```diff
@@ -267,100 +278,123 @@ def release_runahead_tasks(self):
...
-                and runahead_time_limit < self.max_future_offset
-            ):
-                LOG.warning(
-                    f'runahead limit "{runahead_time_limit}" '
-                    'is less than future triggering offset '
-                    f'"{self.max_future_offset}"; workflow may stall.'
-                )
-            self._prev_runahead_base_point = runahead_base_point
+        # TODO: IF FORCE AND NO CHANGE NEED TO CACHE CYCLES!
+        # (max future offset compute)
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/task_pool.py:357]: Right, that was the gnarly one. Fixed.

File: cylc/flow/scheduler.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/scheduler.py:1904]: I take @dwsutherland 's comment to mean we don't need that call at this point.

[the remark this answers is not in the document]
Comment (MetRonnie) [reply cylc/flow/scheduler.py:1904]: Ok maybe not needed 😅 https://github.com/cylc/cylc-flow/pull/4464#discussion_r727302069

File: cylc/flow/task_pool.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MetRonnie) [reply cylc/flow/task_pool.py:311]: I know, just suggesting using a set comprehension instead of for loop

[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/task_pool.py:311]: Oops, so you were!  I'll do that.

**Comment by oliver-sanders** [comment]
> Ubuntu Mono Derivative. You like or no like?

Me likey.

File: cylc/flow/taskdef.py [hunk 8]
```diff
@@ -310,3 +309,58 @@ def is_valid_point(self, point):
...
+        if adjusted:
+            p_next = min(adjusted)
+        return p_next
+
+    def is_parentless(self, point, abs_ok=True):
+        """Return True if no parents at point.
+
+        A task can have parents at some points and not at others.
+
+        Tasks are parentless they have no parents, of if the parents:
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/taskdef.py:344]: Holy crap, that was garbled! :grimacing:

File: tests/functional/spawn-on-demand/16-parents-yes-no/flow.cylc [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+        # Unstall, to see if foo gets triggered by baz at point 2.
+        stall handlers = cylc trigger %(workflow)s//2/baz, \
+                           cylc remove %(workflow)s//1/failer
+[scheduling]
+    cycling mode = integer
+    initial cycle point = 1
+    final cycle point = 4
+    runahead limit = P2  # 1, 2, 3
+    [[xtriggers]]
+        never = xrandom("0")  # never satisfied
```
Comment (MetRonnie) [inline tests/functional/spawn-on-demand/16-parents-yes-no/flow.cylc:19]: Unused?
```suggestion
```

File: tests/functional/runahead/default-future/flow.cylc [hunk 10]
```diff
@@ -3,7 +3,7 @@
     UTC mode = True
     allow implicit tasks = True
     [[events]]
-        inactivity timeout = PT10S
+        inactivity timeout = PT20S
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply tests/functional/runahead/default-future/flow.cylc:6]: I was getting some early aborts due to this, under load, when running many tests at once. Pretty sure not related to this branch.

**Comment by hjoliver** [comment]
(All review comments addressed)

**Review by MetRonnie — APPROVED** [review]
